File: CS/Form1.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using System.IO;
using System.Diagnostics;
using System.Linq;
using System.Linq.Expressions;
using DevExpress.Xpo;
using Northwind;

namespace XpoLinqNorthwindSample {
    public partial class Form1 : Form {
        XpoLogWriter log;
        public Form1() {
            InitializeComponent();
            log = new XpoLogWriter(textBox1);
            System.Diagnostics.Trace.Listeners.Add(log);
        }
        private void Form1_Load(object sender, EventArgs e) {
            // this line is needed to show initial queries produced by XPO 
            // when it connects to the database to retrive an object
            // for the first time after the application is started
            Customer cust = XpoDefault.Session.FindObject<Customer>(null);
        }
        private void btnClear_Click(object sender, EventArgs e) {
            textBox1.Text = "";
        }
        class XpoLogWriter : TraceListener {
            TextBox outputWindow;
            public XpoLogWriter(TextBox outputWindow) {
                this.outputWindow = outputWindow;
            }
            public override void Write(string message) {
                outputWindow.AppendText(message);
            }
            public override void WriteLine(string message) {
                outputWindow.AppendText(message + "\r\n");
            }
        }

        XPQuery<Customer> customers = new XPQuery<Customer>(Session.DefaultSession);
        XPQuery<Order> orders = new XPQuery<Order>(Session.DefaultSession);
        XPQuery<Employee> employees = new XPQuery<Employee>(Session.DefaultSession);

        private void btnSelectWhere_Click(object sender, EventArgs e) {
            log.WriteLine("---------------------------------");
            // simple Select with Where and OrderBy clauses
            var myList = from c in customers
                       where (c.Country == "Germany" && c.ContactTitle == "Sales Representative")
                       orderby c.ContactName
                       select c;
            foreach (Customer cust in myList)
                log.WriteLine(string.Format("{0}\t{1}\t{2}", cust.ContactName, cust.Country, cust.ContactTitle));
        }

        private void btnSelectTop_Click(object sender, EventArgs e) {
            log.WriteLine("---------------------------------");
            // Select Top 5 objects
            var myList = (from o in orders
                        orderby o.ShippedDate descending
                        select o).Take(5);
            foreach (Order order in myList)
                log.WriteLine(string.Format("{0}\t{1}\t{2}", order.OrderID, order.ShippedDate, order.Customer.CompanyName));
        }

        private void btnJoin_Click(object sender, EventArgs e) {
            log.WriteLine("---------------------------------");
            // Group Join customers with an aggregation on their Orders
            var myList = from c in customers
                       join o in orders on c equals o.Customer into oo
                       where oo.Count() >= 1
                       select new { c.CompanyName, OrderCount = oo.Count() };
            foreach (var item in myList)
                log.WriteLine(string.Format("{0}\t{1}", item.CompanyName, item.OrderCount));
        }

        private void btnAggregates_Click(object sender, EventArgs e) {
            log.WriteLine("---------------------------------");
            // an example of aggregated functions (Count and Average)
            var myList = from o in orders
                       select o;
            int count = myList.Count();
            log.WriteLine(string.Format("Orders Row Count: {0}", count));

            decimal avg = myList.Average(x => x.Freight);
            log.WriteLine(string.Format("Orders Average Freight: {0:c2}", avg));
        }

        private void btnGroup_Click(object sender, EventArgs e) {
            log.WriteLine("---------------------------------");
            // Select with Group By
            var myList = from c in customers
                       group c by c.ContactTitle into cc
                       where cc.Count() >= 1
                       select new { Title = cc.Key, Count = cc.Count() };
            foreach (var item in myList)
                log.WriteLine(string.Format("{0}\t{1}", item.Title, item.Count));
        }

        private void btnAny_Click(object sender, EventArgs e) {
            log.WriteLine("---------------------------------");
            bool result = customers.Any(c => c.Country == "Spain");
            log.WriteLine(string.Format("Is there any customer from Spain? {0}", result ? "Yes" : "No"));
            result = customers.Any(c => c.Country == "Monaco");
            log.WriteLine(string.Format("Is there any customer from Monaco? {0}", result ? "Yes" : "No"));
        }
    }
}

File: VB/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports System.IO
Imports System.Diagnostics
Imports System.Linq
Imports System.Linq.Expressions
Imports DevExpress.Xpo
Imports Northwind

Namespace XpoLinqNorthwindSample
	Partial Public Class Form1
		Inherits Form
		Private log As XpoLogWriter
		Public Sub New()
			InitializeComponent()
			log = New XpoLogWriter(textBox1)
			System.Diagnostics.Trace.Listeners.Add(log)
		End Sub


		Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
			' this line is needed to show initial queries produced by XPO 
			' when it connects to the database to retrive an object
			' for the first time after the application is started
			Dim cust As Customer = XpoDefault.Session.FindObject(Of Customer)(Nothing)
		End Sub

		Private Sub btnClear_Click(ByVal sender As Object, ByVal e As EventArgs) Handles simpleButton2.Click
			textBox1.Text = ""
		End Sub

		Private Class XpoLogWriter
			Inherits TraceListener
			Private outputWindow As TextBox
			Public Sub New(ByVal outputWindow As TextBox)
				Me.outputWindow = outputWindow
			End Sub
			Public Overrides Overloads Sub Write(ByVal message As String)
				outputWindow.AppendText(message)
			End Sub
			Public Overrides Overloads Sub WriteLine(ByVal message As String)
				outputWindow.AppendText(message & Constants.vbCrLf)
			End Sub
		End Class

		Private customers As XPQuery(Of Customer) = New XPQuery(Of Customer)(Session.DefaultSession)
		Private orders As XPQuery(Of Order) = New XPQuery(Of Order)(Session.DefaultSession)
		Private employees As XPQuery(Of Employee) = New XPQuery(Of Employee)(Session.DefaultSession)

		Private Sub btnSelectWhere_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnSelectWhere.Click
			log.WriteLine("---------------------------------")
			' simple Select with Where and OrderBy clauses
			Dim myList = _
				From c In customers _
				Where (c.Country = "Germany" AndAlso c.ContactTitle = "Sales Representative") _
				Order By c.ContactName _
				Select c
			For Each cust As Customer In myList
				log.WriteLine(String.Format("{0}" & Constants.vbTab & "{1}" & Constants.vbTab & "{2}", cust.ContactName, cust.Country, cust.ContactTitle))
			Next cust
		End Sub

		Private Sub btnSelectTop_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnSelectTop.Click
			log.WriteLine("---------------------------------")
			' Select Top 5 objects
			Dim myList = ( _
					From o In orders _
					Order By o.ShippedDate Descending _
					Select o).Take(5)
			For Each order As Order In myList
				log.WriteLine(String.Format("{0}" & Constants.vbTab & "{1}" & Constants.vbTab & "{2}", order.OrderID, order.ShippedDate, order.Customer.CompanyName))
			Next order
		End Sub

		Private Sub btnJoin_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnJoin.Click
			log.WriteLine("---------------------------------")
			' Group Join customers with an aggregation on their Orders
			Dim myList = _
				From c In customers _
				Group Join o In orders On c Equals o.Customer Into oo = Group _
				Where oo.Count() >= 1 _
				Select New With {Key c.CompanyName, Key .OrderCount = oo.Count()}
			For Each item In myList
				log.WriteLine(String.Format("{0}" & Constants.vbTab & "{1}", item.CompanyName, item.OrderCount))
			Next item
		End Sub

		Private Sub btnAggregates_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnAggregates.Click
			log.WriteLine("---------------------------------")
			' an example of aggregated functions (Count and Average)
			Dim myList = _
				From o In orders _
				Select o
			Dim count As Integer = myList.Count()
			log.WriteLine(String.Format("Orders Row Count: {0}", count))

			Dim avg As Decimal = myList.Average(Function(x) x.Freight)
			log.WriteLine(String.Format("Orders Average Freight: {0:c2}", avg))
		End Sub

		Private Sub btnGroup_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnGroup.Click
			log.WriteLine("---------------------------------")
			' Select with Group By
			Dim myList = _
				From c In customers _
				Group c By c.ContactTitle Into cc = Group _
				Where cc.Count() >= 1 _
				Select New With {Key .Title = ContactTitle, Key .Count = cc.Count()}
			For Each item In myList
				log.WriteLine(String.Format("{0}" & Constants.vbTab & "{1}", item.Title, item.Count))
			Next item
		End Sub

		Private Sub btnAny_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnAny.Click
			log.WriteLine("---------------------------------")
			Dim result As Boolean = customers.Any(Function(c) c.Country = "Spain")
			log.WriteLine(String.Format("Is there any customer from Spain? {0}",If(result, "Yes", "No")))
			result = customers.Any(Function(c) c.Country = "Monaco")
			log.WriteLine(String.Format("Is there any customer from Monaco? {0}",If(result, "Yes", "No")))
		End Sub
	End Class
End Namespace